cmd/create/machinepool/cmd.go

```diff
@@ -17,255 +17,100 @@ limitations under the License.
+	cmd *cobra.Command, clusterKey string, cluster *cmv1.Cluster,
+	options *mpOpts.CreateMachinepoolUserOptions) error {
+	if cluster.Hypershift().Enabled() {
+		return m.service.CreateNodePools(r, cmd, clusterKey, cluster, options)
+	}
+	return m.service.CreateMachinePool(r, cmd, clusterKey, cluster, options)
+}
 
-	flags.StringVar(&args.RootDiskSize,
-		"disk-size",
-		"",
-		"Root disk size with a suffix like GiB or TiB",
-	)
+// Original function refactored to use the new helper functions
+func CreateMachinepoolRunner(userOptions *mpOpts.CreateMachinepoolUserOptions) rosa.CommandRunner {
+	return func(ctx context.Context, r *rosa.Runtime, cmd *cobra.Command, argv []string) error {
+		var err error
+		options := NewCreateMachinepoolOptions()
+		clusterKey := r.GetClusterKey()
+		options.args = userOptions
+
+		cluster := r.FetchCluster()
+		if err := validateClusterState(cluster, clusterKey); err != nil {
+			return err
+		}
 
-	flags.StringSliceVar(&args.SecurityGroupIds,
-		securitygroups.MachinePoolSecurityGroupFlag,
-		nil,
-		"The additional Security Group IDs to be added to the machine pool. "+
-			"Format should be a comma-separated list.",
-	)
+		val, ok := cluster.Properties()[properties.UseLocalCredentials]
+		useLocalCredentials := ok && val == "true"
 
-	flags.StringVar(&args.NodeDrainGracePeriod,
-		"node-drain-grace-period",
-		"",
-		"You may set a grace period for how long Pod Disruption Budget-protected workloads will be "+
-			"respected when the NodePool is being replaced or upgraded.\nAfter this grace period, all remaining workloads "+
-			"will be forcibly evicted.\n"+
-			"Valid value is from 0 to 1 week (10080 minutes), and the supported units are 'minute|minutes' or "+
-			"'hour|hours'. 0 or empty value means that the NodePool can be drained without any time limitations.\n"+
-			"This flag is only supported for Hosted Control Planes.",
-	)
+		if err := validateLabels(cmd, options.args); err != nil {
+			return err
+		}
 
-	flags.StringSliceVar(
-		&args.Tags,
-		"tags",
-		nil,
-		"Apply user defined tags to all resources created by ROSA in AWS. "+
-			"Tags are comma separated, for example: 'key value, foo bar'",
-	)
+		r.AWSClient, err = aws.NewClient().
+			Region(cluster.Region().ID()).
+			Logger(r.Logger).
+			UseLocalCredentials(useLocalCredentials).
+			Build()
+		if err != nil {
+			return fmt.Errorf("Failed to create awsClient: %s", err)
+		}
+		service := NewCreateMachinePool(CreateMachinePoolSpec{
+			Service: machinepool.NewMachinePoolService(),
+		})
 
-	interactive.AddFlag(flags)
-	output.AddFlag(Cmd)
+		return service.createMachinePoolBasedOnClusterType(r, cmd, clusterKey, cluster, options.Machinepool())
+	}
 }
 
-func run(cmd *cobra.Command, _ []string) {
-	r := rosa.NewRuntime().WithOCM()
-	defer r.Cleanup()
-
-	clusterKey := r.GetClusterKey()
-
-	cluster := r.FetchCluster()
+// Validate the cluster's state is ready
+func validateClusterState(cluster *cmv1.Cluster, clusterKey string) error {
```

Comment:
@den-rgb Same comment here about moving this into the service. I think the service should check to see if the cluster is in `ready` state and return the error if not.

Reply:
I don't know if this would be necessary to move since its only really being used by the runner?

Reply:
yeah i would plus one what Rob said and move to the service. Keep the runner as dumb as possible, the runner really should not be making decisions. Only parsing/binding args and passing to the service. Leave the service make decisions